### ISSUE TITLE
Write integration tests

### ISSUE THREAD

**Issue by zsoltk** [zsoltk/paperwork#28]
The idea is to create another Android module in the project specifically for these tests. Using the paperwork-example module is not ok, as it grabs its dependencies from Maven Central instead of local code (and it's good that way, so people can copy-paste that if they want).
1. Create a `paperwork` config block in `build.gradle` using all the helpers
2. Create an `Activity`, with `TextView`s bound to each generated value
3. Write test(s) to test the values, preferably with Espresso

@vanniktech any thoughts on this?

**Comment by zsoltk** [comment]
It doesn't seem possible to add a project module dependency to the buildscript block (circular dependency). But then how to include the plugin from its module for the integration tests?

@vanniktech do you have any ideas?

**Comment by vanniktech** [comment]
Hmm sorry for answering that late and no I don't have an idea. 

We could test the Gradle Plugin though that the moment the evaluation phase of the project has ended we do generate the file. That's probably as close to 'integration tests' we can get.

**Comment by zsoltk** [comment]
I wonder if we could first generate a .jar from the plugin and copy that over to the integration test module, as jar dependencies can be used, and only then run the tests (all this in an automated way, somehow). Or copy the contents of the plugin & extension over to either buildSrc or an included gradle file. I know these ideas sound horribly hack-ish, but there's a problem to be solved here :) What do you think?

**Comment by vanniktech** [comment]
Hmm if we can find a way sure, though it does require a lot of work :/

**Comment by zsoltk** [comment]
I worked something out. Publishing soon.

**Comment by vanniktech** [comment]
Nice looking forward to it :+1:

**Comment by zsoltk** [comment]
Ok, so here's a summary:
- The solution is to copy the two .groovy files from the plugin to their respective directories under `buildSrc` in the root dir, as any class that can be found there can and will be included when building the integration module (or any other module)
- I created the directory structure under `buildSrc` with a `.gitignore` file inside, ignoring any .groovy files there, so we don't have to deal with them
- I created a `refresh-integration-dependencies.sh` file under `scripts`, which copies over said files to `buildSrc`
- Unfortunately, `buildSrc` is processed before anything else by gradle, even init hooks, so if the classes are not there _already_ before running any gradle task, the plugin will not be found, and the build will fail
- Since I didn't want to anyone checking out the code to have to mess around with copying the files / calling `refresh-integration-dependencies.sh` by hand, the script is now included in `gradlew` as the very first thing to run. This way the files are guaranteed to always be there, and be the newest version automacially.
- The only problem is that `paperwork-example` uses the plugin from maven central, and will clash with the one found in `buildSrc`. First I tried using scripts to work around this problem (automatically deleting the files from `buildSrc`), but I realized it's not worth all the hassle. As a result, the example module is now disabled in `settings.gradle` and will stay only as a read-only code example. I wouldn't think it's really all that important to build it, but if someone wants to, they can do it by disabling the other modules and removing the actual `cp` operation from the shell script. Running integration tests easily seems a lot more important.

As for the actual testing:
- Run `scripts/run-tests.sh` (this is referenced in `.travis.yml` as well)
- I had to increase the timestamp thresholds in plugin / integration tests to 10 seconds / 10 minutes respectively, as travis can be quite slow sometimes, and otherwise they will fail all too often
- I had to replace a meaningful test for buildTime2 (using no timezone when generating) with checking only that "some" text is present, as the build environment and the emulator might have totally different timezones. I tried setting the timezone of the emulator through adb shell, but its result is not immediate (like setting the emulator time, which I'm already doing), and would only take place after the next boot cycle. Well maybe we could test for something that's not dependent on the timezone. Era maybe? ("AD")
- Sometimes Travis fail _all_ integration tests with an error `java.lang.RuntimeException: Waited for the root of the view hierarchy to have window focus and not be requesting layout for over 10 seconds` (see [this build](https://travis-ci.org/zsoltk/paperwork/builds/101726523), and I'm not sure what to do with it, as it happens randomly (but all too often to ignore it). Resources on the web say this might be because the emulator is not unlocked. In theory, it is (by sending keyevent 82), but maybe it doesn't always receive it? I wonder if you could help with this.

The code is pushed to the `integration-test-module` branch. I will wait for your input and any changes you might want to commit before merging it into `master`.

When done, I'll release version 1.2 and call it a day.

**Comment by vanniktech** [comment]
Sounds splendid.

Have a look [here](https://github.com/JakeWharton/RxBinding/blob/master/testing-utils/src/main/java/com/jakewharton/rxbinding/RxBindingTestRunner.java). If we also have a `TestRunner` maybe we can also unlock the device programmatically.

As for the pushed branch, can you create a PR for that and I'll review it from there?

**Comment by zsoltk** [comment]
I already have a custom TestRunner like the one you linked, and it didn't help either.

I though you might wanted to add your own PR-s to this branch maybe? Comparison [here](https://github.com/zsoltk/paperwork/compare/integration-test-module). If you still want me doing the PR, tell me.

**Comment by vanniktech** [comment]
Yeah it'd be nice to have a PR for that. Then I could comment on it directly and also once everything is done you'd just rebase and squash everything into one commit.

If I wanted to add something I could also cherry pick your PR from your repository.

**Comment by zsoltk** [comment]
See #32.
